<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function evaluates the presence of cancer in a sample by combining the cancerous signal across a catalogue of candidate mutations."><title>Call cancer from read positions in bam-file — dreams_cc • dreams</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Call cancer from read positions in bam-file — dreams_cc"><meta property="og:description" content="This function evaluates the presence of cancer in a sample by combining the cancerous signal across a catalogue of candidate mutations."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dreams</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-solid fa-cloud-moon"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/dreams.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/JakobSkouPedersenLab/dreams/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Call cancer from read positions in bam-file</h1>
      <small class="dont-index">Source: <a href="https://github.com/JakobSkouPedersenLab/dreams/blob/HEAD/R/call_cancer.R" class="external-link"><code>R/call_cancer.R</code></a></small>
      <div class="d-none name"><code>dreams_cc.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function evaluates the presence of cancer in a sample by combining the cancerous signal across a catalogue of candidate mutations.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">dreams_cc</span><span class="op">(</span>
  <span class="va">mutations_df</span>,
  <span class="va">bam_file_path</span>,
  <span class="va">reference_path</span>,
  <span class="va">model</span>,
  <span class="va">beta</span>,
  alpha <span class="op">=</span> <span class="fl">0.05</span>,
  calculate_confidence_intervals <span class="op">=</span> <span class="cn">FALSE</span>,
  use_turboem <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>mutations_df</dt>
<dd><p>A <code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame()</a></code> with candidate mutations (SNVs) (chromosome, positions, reference and alternative)</p></dd>


<dt>bam_file_path</dt>
<dd><p>Path to .BAM-file</p></dd>


<dt>reference_path</dt>
<dd><p>Path to reference genome e.g. FASTA-file.</p></dd>


<dt>model</dt>
<dd><p>A dreams model. See <code><a href="train_dreams_model.html">train_dreams_model()</a></code>.</p></dd>


<dt>beta</dt>
<dd><p>Down sampling parameter from <code><a href="get_training_data.html">get_training_data()</a></code> for correcting the error-rates from the DREAMS model.</p></dd>


<dt>alpha</dt>
<dd><p>Alpha-level used for testing and confidence intervals. Default is 0.05.</p></dd>


<dt>calculate_confidence_intervals</dt>
<dd><p>Logical. Should confidence intervals be calculated? Default is FALSE.</p></dd>


<dt>use_turboem</dt>
<dd><p>Logical. Should <code><a href="https://rdrr.io/pkg/turboEM/man/turboem.html" class="external-link">turboEM::turboem()</a></code> be used for EM algorithm? Default is TRUE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code><a href="https://rdrr.io/r/base/list.html" class="external-link">list()</a></code> with:</p><ul><li><p><strong>cancer_info</strong> A <code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame()</a></code> with results for cancer calling across all mutations:</p><dl><dt>tf_est</dt>
<dd><p>The estiamted tumor fraction (allele fraction).</p></dd>

<dt>tf_min, tf_max</dt>
<dd><p>The confidence interval of <code>tf_est</code>.</p></dd>

<dt>r_est, est_mutations_present</dt>
<dd><p>The estiamted fraction/number of candidate mutations present in the sample.</p></dd>

<dt>r_min, r_max</dt>
<dd><p>The confidence interval of <code>r_est</code>.</p></dd>

<dt>mutations_tested</dt>
<dd><p>Number of candidate mutations tested.</p></dd>

<dt>total_coverage, total_count</dt>
<dd><p>Total count and coverage across all mutations (only reference and alternative allele(s).</p></dd>

<dt>mutations_tested</dt>
<dd><p>Number of candidate mutations tested.</p></dd>

<dt>EM_converged</dt>
<dd><p>If the EM algorithm converged.</p></dd>

<dt>EM_steps, fpeval, objfeval</dt>
<dd><p>Number of steps and function evaluations by the EM algorithm.</p></dd>

<dt>ll_0, ll_A</dt>
<dd><p>The value of the log-likelihood function under the null (tf=0) and alternative (tf&gt;0) hypothesis.</p></dd>

<dt>Q_val, df, p_val</dt>
<dd><p>The chisq test statistic, degrees of freedom and p-value of the statistical test.</p></dd>

<dt>cancer_detected</dt>
<dd><p>Whether cancer was detected at the supplied alpha level.</p></dd>


</dl></li>
<li><p><strong>mutation_info</strong> A <code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame()</a></code> with information about the individual mutations:</p><dl><dt>chr, genomic_pos</dt>
<dd><p>The genomic position of the mutation.</p></dd>

<dt>ref, alt</dt>
<dd><p>The reference and alternative allele.</p></dd>

<dt>P_mut_is_present</dt>
<dd><p>The estimated probability the mutation is present in the sample.</p></dd>

<dt>exp_count</dt>
<dd><p>The expected count of the alternative allele under the error (null) model.</p></dd>

<dt>count</dt>
<dd><p>The count of the alternative allele.</p></dd>

<dt>coverage</dt>
<dd><p>The coverage used by the model (only referenceredas with and alternative allele).</p></dd>

<dt>obs_freq</dt>
<dd><p>The observed frequency of the alternative allele.</p></dd>


</dl></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="call_mutations.html">call_mutations()</a></code>, <code><a href="train_dreams_model.html">train_dreams_model()</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Mikkel Hovden Christensen, Simon Opstrup Drue.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

    </footer></div>

  

  

  </body></html>

