<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Cancer Detection in Indels Using DREAMS — dreams_cc_indels • dreams</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Cancer Detection in Indels Using DREAMS — dreams_cc_indels"><meta property="og:description" content="This function performs cancer detection based on indel information from a BAM file using a predefined list of mutations.
It processes the mutations data and uses the provided model to make cancer calls."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dreams</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/JakobSkouPedersenLab/dreams/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Cancer Detection in Indels Using DREAMS</h1>
    <small class="dont-index">Source: <a href="https://github.com/JakobSkouPedersenLab/dreams/blob/HEAD/R/call_cancer_indels.R" class="external-link"><code>R/call_cancer_indels.R</code></a></small>
    <div class="hidden name"><code>dreams_cc_indels.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs cancer detection based on indel information from a BAM file using a predefined list of mutations.
It processes the mutations data and uses the provided model to make cancer calls.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">dreams_cc_indels</span><span class="op">(</span></span>
<span>  <span class="va">mutations_df</span>,</span>
<span>  <span class="va">bam_file_path</span>,</span>
<span>  <span class="va">reference_path</span>,</span>
<span>  <span class="va">model</span>,</span>
<span>  <span class="va">model_indels</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  calculate_confidence_intervals <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use_turboem <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>mutations_df</dt>
<dd><p>A dataframe containing the list of mutations to be analyzed.</p></dd>


<dt>bam_file_path</dt>
<dd><p>Path to the BAM file containing sequencing data.</p></dd>


<dt>reference_path</dt>
<dd><p>Path to the reference genome file, typically in FASTA format.</p></dd>


<dt>model</dt>
<dd><p>The model to be used for cancer detection.</p></dd>


<dt>model_indels</dt>
<dd><p>The model to be used for calling mutations for indels.</p></dd>


<dt>alpha</dt>
<dd><p>Significance level for statistical testing, default is 0.05.</p></dd>


<dt>calculate_confidence_intervals</dt>
<dd><p>Logical flag indicating whether to calculate confidence intervals, default is FALSE.</p></dd>


<dt>use_turboem</dt>
<dd><p>Logical flag indicating whether to use the turboEM algorithm, default is TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code><a href="https://rdrr.io/r/base/list.html" class="external-link">list()</a></code> with:</p><ul><li><p><strong>cancer_info</strong> A <code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame()</a></code> with results for cancer calling across all mutations:</p><dl><dt>tf_est</dt>
<dd><p>The estiamted tumor fraction (allele fraction).</p></dd>

<dt>tf_min, tf_max</dt>
<dd><p>The confidence interval of <code>tf_est</code>.</p></dd>

<dt>r_est, est_mutations_present</dt>
<dd><p>The estiamted fraction/number of candidate mutations present in the sample.</p></dd>

<dt>r_min, r_max</dt>
<dd><p>The confidence interval of <code>r_est</code>.</p></dd>

<dt>mutations_tested</dt>
<dd><p>Number of candidate mutations tested.</p></dd>

<dt>total_coverage, total_count</dt>
<dd><p>Total count and coverage across all mutations (only reference and alternative allele(s).</p></dd>

<dt>mutations_tested</dt>
<dd><p>Number of candidate mutations tested.</p></dd>

<dt>EM_converged</dt>
<dd><p>If the EM algorithm converged.</p></dd>

<dt>EM_steps, fpeval, objfeval</dt>
<dd><p>Number of steps and function evaluations by the EM algorithm.</p></dd>

<dt>ll_0, ll_A</dt>
<dd><p>The value of the log-likelihood function under the null (tf=0) and alternative (tf&gt;0) hypothesis.</p></dd>

<dt>Q_val, df, p_val</dt>
<dd><p>The chisq test statistic, degrees of freedom and p-value of the statistical test.</p></dd>

<dt>cancer_detected</dt>
<dd><p>Whether cancer was detected at the supplied alpha level.</p></dd>


</dl></li>
<li><p><strong>mutation_info</strong> A <code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame()</a></code> with information about the individual mutations:</p><dl><dt>chr, genomic_pos</dt>
<dd><p>The genomic position of the mutation.</p></dd>

<dt>ref, alt</dt>
<dd><p>The reference and alternative allele.</p></dd>

<dt>P_mut_is_present</dt>
<dd><p>The estimated probability the mutation is present in the sample.</p></dd>

<dt>exp_count</dt>
<dd><p>The expected count of the alternative allele under the error (null) model.</p></dd>

<dt>count</dt>
<dd><p>The count of the alternative allele.</p></dd>

<dt>coverage</dt>
<dd><p>The coverage used by the model (only referenceredas with and alternative allele).</p></dd>

<dt>obs_freq</dt>
<dd><p>The observed frequency of the alternative allele.</p></dd>


</dl></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="call_mutations.html">call_mutations()</a></code>, <code><a href="train_dreams_model.html">train_dreams_model()</a></code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Mikkel Hovden Christensen, Simon Opstrup Drue.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

